### ISSUE TITLE
FFMPEG5 breaks audio transcode

### ISSUE THREAD

**Issue by smf007** [xbmc/xbmc#22425]
### Describe the bug
Here is a clear and concise description of what the problem is:

Using an android build of nightly, 2023 Jan 7 [8fdbdb18] i.e. Pull #21248, the audio transcode no longer produces an audio output stream.

## Expected Behavior
Here is a clear and concise description of what was expected to happen:

The 5.1 AAC audio should have been transcoded to 5.1 AC3.

## Actual Behavior

An audio sink/render error occurs and no audio output stream results.  If left long enough, Kodi crashes.

### To Reproduce

1. Install the latest nightly Kodi
2. Setup audo transcode
3. Play a video that has an audio track that would require transcoding

### Debuglog

The debuglog can be found here:
https://paste.kodi.tv/igiyerusex.kodi

### Screenshots 
Here are some links or screenshots to help explain the problem:

## Additional context or screenshots (if appropriate)
Here is some additional context or explanation that might help:

### Your Environment
Used Operating system:

 - [x] Android
 - [ ] iOS
 - [ ] tvOS
 - [ ] Linux
 - [ ] macOS
 - [ ] Windows
 - [ ] Windows UWP

 - Operating system version/name: Android TV 10
 - Kodi version: Master, nightly, 2023 Jan 7 [8fdbdb18] - Pull #21248

*note: Once the issue is made we require you to update it with new information or Kodi versions should that be required.
Team Kodi will consider your problem report however, we will not make any promises the problem will be solved.*

**Comment by lrusak** [comment]
Hmmm, all I can see in the log is this:
```
2023-01-08 21:44:48.707 T:19345   error <general>: CDVDAudio::AddPacketsRenderer - timeout adding data to renderer
```

**Comment by basilgello** [comment]
@smf007 Can you post screenshots of relevant transcoding settings so I can reproduce the issue?

**Comment by smf007** [comment]
Settings are shown below.  These settings works fine with the Jan 5th build, prior to the pull.

![A059BC19-B9FB-4EFB-913D-DFE60E4C195B](https://user-images.githubusercontent.com/7517171/211247728-0541f11d-ff9a-4372-9bb9-a861c89703f8.jpeg)
![DA8D8F6B-2B0A-4C59-8BA7-CEF580C46BF8](https://user-images.githubusercontent.com/7517171/211247736-a3a05325-79b3-4674-ae99-4b8cbb20ea1c.jpeg)

**Comment by smf007** [comment]
I don't really know the code, but should https://github.com/xbmc/xbmc/blob/master/xbmc/cores/AudioEngine/Encoders/AEEncoderFFmpeg.cpp#L302 be a while loop?

**Comment by fritsch** [comment]
Yes. Should be a while loop. It's the send / receive API, used in polling mode.

**Comment by smf007** [comment]
The previous code, using the older API, doesn't have that loop.  My question then is, how is the encoded packet then being transmitted on?  Would it not sit there in the loop until it finished encoding the whole audio stream?

**Comment by basilgello** [comment]
No, it loops for receiver and the receiver puts out either one or two packets. At least, should work that way :)

**Comment by smf007** [comment]
So, avcodec_send_frame loads up a frame of audio.  avcodec_receive_packet spits back encoded packets.  When avcodec_receive_packet has nothing further to give, it should return a value to break the loop.

In this [stack overflow example](https://stackoverflow.com/questions/59470927/how-to-replace-avcodec-encode-audio2-avcodec-encode-video2-with-avcodec-send), they are obviously writing out to file, but they are using that packet before unreferencing it.  If I'm mistaken, in Kodi now I don't see anything happening with the encoded packets.  What does variable pkt has after the while loop?  Wouldn't it be null?

**Comment by basilgello** [comment]
Interesting point. I have just checked the extradata commit of the same chain and the pktRef received is used. Let me check the ffmpeg docs.

**Comment by basilgello** [comment]
The https://ffmpeg.org/doxygen/4.0/group__lavc__encdec.html states tnat "Providing user-allocated buffers for avcodec_receive_packet() is not possible." but in trunk revision of the same page the whole example is omitted. It looks like we need to copy contents of `pkt->data` into `out` checking the size of all packets against buffer size.

**Comment by smf007** [comment]
I think the code currently expects pkt to be send back to the calling routine.
https://github.com/xbmc/xbmc/blob/master/xbmc/cores/AudioEngine/Engines/ActiveAE/ActiveAE.cpp#L2277
That would limit to only returning 1 packet though.

**Comment by fritsch** [comment]
Yes, that's the case: https://github.com/xbmc/xbmc/blob/master/xbmc/guilib/FFmpegImage.cpp#L702  - see also FFmpegPicture for the same. We were used to not allocate new memory for the output before, but this is now a change in the new API (several years old though), has hit us in ffmpeg picture before as well, via: https://github.com/xbmc/xbmc/commit/4b8572a13810c3cc9ad88277c379d0f29770afd5

**Comment by fritsch** [comment]
And while it is easy for the picture use-case: One input, one output. The AC3 encoder might output more pkts from one frame, therefore besides the missing copy, the while loop, though not being wrong, might only (after memcpy is fixed) deliver the last package before EAGAIN / EOF happens.

EAGAIN btw. is not an error, it just means: nothing here at this point, but might come later ... this was iirc invented when certain encoders / decoders work asynchroneously in the background to not block callers infinitely.

Which brings me to the discussion we had some years ago about our MP3 Encoder inside kodi, that the new API should be used in a way, that it brings benefit, not to try "wrapping old behaviour".

**Comment by basilgello** [comment]
I'd better engage into automated functional testing discussion - it is definitely doable to capture audio, video and logs to catch at least "no sound" issues properly. Maybe we should think of it as a GSoC project?

**Comment by basilgello** [comment]
@smf007 https://mirrors.kodi.tv/test-builds/android/arm/kodi-20230115-ddd8da9a-ffmpeg5.0-Nexus-27912-armeabi-v7a.apk can you test please?

**Comment by smf007** [comment]
Unfortunately that doesn't work.  My amp though shows 2 channels instead of none, but it's still not receiving anything.  Also no change in that the video works fine for a few seconds, then gets sluggish and eventually freezes.  Kodi then gets terminated.
https://paste.kodi.tv/firafuxobu.kodi
